cloud_functions/test/flagCryptominer-error-tests.js

```diff
@@ -0,0 +1,94 @@
+// These tests cover various error cases for the flagCryptominer cloud function. Errors are exposed
+// via Cloud Logging Error Reporting.
+// See https://cloud.google.com/functions/docs/monitoring/error-reporting for details.
+//
+// These events are handled asynchronously from the processes emitting the events. Errors can not be
+// reported directly back to callers. Therefore, it is highly valuable to make sure that unexpected
+// conditions will be detected and reported.
+//
+// Uncaught exceptions will log errors, but will also incur a cold start on future invocations.
+// Instead, we use console.error, so we need to stub that function for these tests.
+//
+// IMPORTANT: Since we're stubbing a global function (console.error), all of these tests must run
+// in serial, not parallel as is Ava's default.
+// (For future exploration: https://github.com/sinonjs/sinon-test)
+//
+// Based on https://cloud.google.com/functions/docs/testing/test-background#unit_tests
+
+const test = require('ava')
+const flagCryptominer = require('../flagCryptominer')
+const sinon = require('sinon')
+const utils = require('./test-utils')
+
+
+test.beforeEach(t => {
+  sinon.stub(console, 'error')
+})
+
+test.afterEach.always(t => {
+  console.error.restore()
+})
+
+
+// The function should do nothing if the cryptominer attribute isn't set. This could indicate a
+// misconfigured Pub/Sub topic subscription. In this case, the function should do nothing in order
+// to avoid doing the wrong thing.
+test.serial('reject non-cryptominer messages', async t => {
+  const mixpanel = utils.makeStubMixpanel()
+  const message = {
+    data: utils.encodeMessageData({ userSubjectId: '123' })
+  }
+
+  await flagCryptominer.run(mixpanel, message)
+
+  checkConsoleError(t, /unexpected message/i)
+  t.assert(mixpanel.people.set.notCalled)
+})
+
+test.serial('data is not JSON', async t => {
+  const mixpanel = utils.makeStubMixpanel()
```

Comment:
can you make the mixpanel stub in the `beforeEach` to DRY things out?

Reply:
I think that would require `mixpanel` to be global and some kind of reset to be done in `afterEach`. That wouldn't be the end of the world, but I'd rather keep things out of global scope whenever possible.

I'd like for `console` to be the only exception causing these to have to run in serial rather than allowing the fact that these have to run in serial to justify adding more global state.